File: packages/rust/src/types/environment.rs
//! Environment types (Rust-native)

use serde::{Deserialize, Serialize};
use std::collections::HashMap;

use super::primitives::UUID;

/// Represents a world (server, guild, or top-level container).
#[derive(Clone, Debug, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct World {
    pub id: UUID,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub name: Option<String>,
    pub agent_id: UUID,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub message_server_id: Option<UUID>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub metadata: Option<WorldMetadata>,
}

/// World metadata.
#[derive(Clone, Debug, Default, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct WorldMetadata {
    #[serde(skip_serializing_if = "Option::is_none")]
    pub ownership: Option<WorldOwnership>,
    #[serde(default, skip_serializing_if = "HashMap::is_empty")]
    pub roles: HashMap<String, String>,
    #[serde(default, skip_serializing_if = "HashMap::is_empty")]
    pub extra: HashMap<String, serde_json::Value>,
}

/// World ownership metadata.
#[derive(Clone, Debug, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct WorldOwnership {
    pub owner_id: UUID,
}

/// Room metadata (dynamic key-value pairs).
#[derive(Clone, Debug, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct RoomMetadata {
    #[serde(default, skip_serializing_if = "HashMap::is_empty")]
    pub values: HashMap<String, serde_json::Value>,
}

/// Represents a room (channel, chat, or conversation container).
#[derive(Clone, Debug, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct Room {
    pub id: UUID,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub name: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub agent_id: Option<UUID>,
    pub source: String,
    #[serde(rename = "type")]
    pub room_type: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub channel_id: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub message_server_id: Option<UUID>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub world_id: Option<UUID>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub metadata: Option<RoomMetadata>,
}

/// Room participant with entity details.
#[derive(Clone, Debug, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct Participant {
    pub id: UUID,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub entity: Option<Entity>,
}

/// Entity component - extensible data attached to entities.
#[derive(Clone, Debug, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct Component {
    pub id: UUID,
    pub entity_id: UUID,
    pub agent_id: UUID,
    pub room_id: UUID,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub world_id: Option<UUID>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub source_entity_id: Option<UUID>,
    #[serde(rename = "type")]
    pub component_type: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub created_at: Option<i64>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub data: Option<serde_json::Value>,
}

/// Represents an entity (user, agent, or other actor).
#[derive(Clone, Debug, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct Entity {
    #[serde(skip_serializing_if = "Option::is_none")]
    pub id: Option<UUID>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub names: Option<Vec<String>>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub metadata: Option<serde_json::Value>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub agent_id: Option<UUID>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub components: Option<Vec<Component>>,
}

/// Represents a relationship between entities.
#[derive(Clone, Debug, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct Relationship {
    pub id: UUID,
    pub source_entity_id: UUID,
    pub target_entity_id: UUID,
    pub agent_id: UUID,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub tags: Option<Vec<String>>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub metadata: Option<serde_json::Value>,
}

File: packages/typescript/src/types/environment.ts
import type { ChannelType, Metadata } from "./primitives";
import type {
  Component as ProtoComponent,
  Entity as ProtoEntity,
  Participant as ProtoParticipant,
  Relationship as ProtoRelationship,
  Room as ProtoRoom,
  World as ProtoWorld,
  WorldMetadata as ProtoWorldMetadata,
} from "./proto.js";
import type { WorldSettings } from "./settings";

export type TimestampValue = number;

export interface Component
  extends Omit<
    ProtoComponent,
    "$typeName" | "$unknown" | "createdAt" | "data"
  > {
  createdAt: TimestampValue;
  data?: Metadata;
}

/**
 * Represents a user account
 */
export interface Entity
  extends Omit<
    ProtoEntity,
    "$typeName" | "$unknown" | "metadata" | "components"
  > {
  metadata?: Metadata;
  components?: Component[];
}

/**
 * Defines roles within a system, typically for access control or permissions, often within a `World`.
 * - `OWNER`: Represents the highest level of control, typically the creator or primary administrator.
 * - `ADMIN`: Represents administrative privileges, usually a subset of owner capabilities.
 * - `NONE`: Indicates no specific role or default, minimal permissions.
 * These roles are often used in `World.metadata.roles` to assign roles to entities.
 */
export const Role = {
  OWNER: "OWNER",
  ADMIN: "ADMIN",
  NONE: "NONE",
} as const;

export type Role = (typeof Role)[keyof typeof Role];

export interface WorldOwnership {
  ownerId: string;
}

export interface WorldMetadata
  extends Omit<
    ProtoWorldMetadata,
    "$typeName" | "$unknown" | "roles" | "extra" | "ownership"
  > {
  type?: string;
  description?: string;
  ownership?: WorldOwnership;
  roles?: Record<string, Role>;
  extra?: Metadata;
  settings?: WorldSettings;
}

export interface World
  extends Omit<ProtoWorld, "$typeName" | "$unknown" | "metadata"> {
  metadata?: WorldMetadata;
}

export interface Room
  extends Omit<ProtoRoom, "$typeName" | "$unknown" | "type" | "metadata"> {
  type: ChannelType;
  metadata?: Metadata;
}

export type RoomMetadata = Metadata;

/**
 * Room participant with account details
 */
export interface Participant
  extends Omit<ProtoParticipant, "$typeName" | "$unknown" | "entity"> {
  entity: Entity;
}

/**
 * Represents a relationship between users
 */
export interface Relationship
  extends Omit<ProtoRelationship, "$typeName" | "$unknown" | "metadata"> {
  metadata?: Metadata;
}
